### ISSUE TITLE
Idea for IOV package

### ISSUE THREAD

**Issue by tunnell** [tunnell/siovd#1]
We can use a pandas DataFrame to store global conditions / configuration / correction information for an experiment.  Have the data for the DataFrame be stored as CSV or better yet Google doc so widely accessible:

https://stackoverflow.com/questions/24063629/loading-a-generic-google-spreadsheet-in-pandas

In the dataframe:
Row: date
Column: global version

(Bonus points: multindex of date, correction name?)

Global version means a unique set of correction maps and variables (electron lifetime, liquid level height, etc) when running e.g. strax.   Therefore, analysts can say "I used strax v5 and corrections v4.3".  

Furthermore, global versions will be split so:

* Online (corrections used at initial processing)
* Offline
  * v1.0: uses v2 of correction Y, v0 of map X, etc
  * v1.0: uses v2 of correction Y, v2 of map X, etc

Where each of the list above is a column in dataframe.  

Each element in the DataFrame will be a reference to something that can be downloaded.  For example, "v2 of correction Y" would be something unique ID "correction_Y_v2".  This string is stored in table.  This string can be given to another function to fetch the files corresponding to this unique ID.  These files can live on Github, MongoDB GridFS, HTTP, or S3.

**Comment by coderdj** [comment]
Good idea. Bit nervous about google doc but would be convenient, and also fine as long as we weren't dependent on it being a google doc (since google likes to change things without notice).

Still will have some issues to solve with corrections that change over time and have to be extrapolated into the future, i.e. electron lifetime and gains. (OK we hope not gains but I guess we should plan for it)

Possible solutions would be evaluating some time-dependent function to fill this table at either run start or initial processing by whatever cax-like daemon handles that stuff. Or pre-compute using time, not run number, as binning. Could also tie-in our online calibrations like EL analysis and LED calibration to automatically use the most recent value determined valid from these.

**Comment by tunnell** [comment]
It could also be a table in a lonely Github repo.

Now that I think of it, I think extending the time into the future is also how other experiments do it. 

So if ATLAS sets variable X to have value Y from now, this "interval of validity" or rather time range is today -> cold death of Universe.  Therefore, we can do the following: if use day index, then set value of electron lifetime for today, extrapolate to tomorrow, and day after, etc.  So we can extrapolate from today to months from now.  Then in a week we redo this for all values only in future.  This is for the 'online'.

I'm not sure how you handle the offline versions if these extrapolations change weekly...

Also, if we index by day, then you can't correct an EL drop happening at noon and loose 12 hours of data.  I guess you can get around the day resolution by just adding an extra column at the time of interest.  Therefore, the spacing isn't always day.

**Comment by JelleAalbers** [comment]
Just to summarize how I did it / planned it for strax at the moment:

A correction setting is like any other config option: it is fully tracked with every piece of data that depends on it. Users can request data with any correction setting, so when we talk about tracking corrections, what we're tracking is just the time-dependent defaults of those options.

Currently these defaults can be specified by range of validity, much like in hax (https://github.com/XENON1T/hax/blob/master/hax/hax.ini#L113), though it is decentralized: each plugin specifies only it's own options. See e.g. https://github.com/AxFoundation/strax/blob/master/strax/xenon/plugins.py#L496. The range is specified by the run names. Note that our XENON1T run names, including underscores, are valid integer literals in python 3.6 :-).

You'll want some centralized global master versioning system too. This is what Contexts are designed for. The analysis coordinators can define a Context with all the correct options set and plugins registered, then  analysts load that (and customize it as they want/dare). So data loading would look like:
```
import strax_nt
sc = strax_nt.get_context('sr1_v2')
df = sc.get_df(run_name, 'event_info')
```

As for the mechanics of loading correction files: the settings themselves should just be filenames without paths such as `XENONnT_fdc_3d_v12_final_really.json.gz`. The plugins (if they are called up for processing) invoke a common `get_resource` method that downloads the map and caches it locally. Currently this loads it from github, but it could also be the runs db, or choose between them. (This isn't quite how it's implemented; now the setting is a full URL, but speaking to Chris this is probably not sustainable)

I haven't implemented a solution for fast-changing non-file-based things like electron lifetime and gains; these change too often to track them in a file in a repository. For elife it seems easiest to just store a single number in the run-level metadata (i.e. the runs db). For gains that could work too; if we're too stingy to store 500 numbers per run, we could store a version number of a gain model and download the corresponding file.
